Rebase mick-p1982's PR #13

nanoant opened the pull request:
This is rebase of @mick-p1982 PR #13, so it can be merged into the master.

nanoant commented on CMakePCHCompiler.cmake:
```diff
@@ -224,6 +204,12 @@ macro(__define_pch_compiler lang)
 			)
+		#2017: Enable MSVC2010 and forward to work.
+		#With /Zi Microsoft.Cpp.Win32.targets deletes the PCH file.
+		string(APPEND CMAKE_${lang}_FLAGS_DEBUG " /Z7")
```
Do we really need such a drastic solution for MSVC? I understand this effectively turns off PDB generation and puts debug information in .obj files itself. This can be a really showstopper for many users, and has to be carefully documented in the README.

mick-p1982 replied:
Yeah, I would document it. As far do you need it? Well, yes, for MSVC2010 forward, which is all relevant version of MSVC. I'd be curious to know what the downside is, but if you can debug, and see quick build times, I don't see a problem. In any case, yes, it's not happening any other way.

nanoant commented:
@mick-p1982 I have reviewed your changed more thoroughly this time. While first "quoting", `PRIVATE` commits seems to be obvious and I am ready to merge them, the latter ones attempting to solve MSVC PCH removal, and your final last commit's solution of `/Z7` looks a bit drastic (and limiting) to me, as long as I understand correctly that this effectively turns of PDB generation in debug builds when using PCH with this module.
I wish to look into the issue, can you provide me more information about this? Where & how this was exactly reported to Microsoft if it is unintended behavior.

mick-p1982 commented:
I want the module to be useful, or would just assume the changes not be implemented. There is a long discussion on the Kitware about the z7 and other issues, that I don't want to think about further. It's a miracle that it works at all. The only thing that matters is if you can build with PCH or not. Nothing else matters.

mick-p1982 commented:
@nanoant The short answer is, with PDB Visual Studio will build the PCH target, and then delete its output, and then fail to build the main target. Obviously, that's not viable and never will be.

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -78,27 +90,29 @@ function(target_precompiled_header) # target [...] header
 					PROPERTIES
 					LANGUAGE ${lang}
 					COMPILE_FLAGS ${flags}
+					#COMPILE_PDB_OUTPUT_DIRECTORY ${pdb_dir}) ##/Fd\"${pdb_dir}\\\\\"
```
It's not in my working-copy. I don't really have time for this. GitHub is hiding these anyway (says Hide outdated) and only found them out of curiosity.

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -78,27 +90,29 @@ function(target_precompiled_header) # target [...] header
+
+				#Note, this adds debug/release/etc. onto the end.
+				set_target_properties(${pch_target} PROPERTIES
+					COMPILE_PDB_OUTPUT_DIRECTORY ${pdb_dir}) ##/Fd\"${pdb_dir}\\\\\"
```
PDB is abandoned. It's not viable. It's an MSVC feature. And it doesn't work with multi-target CMake. Enough said.

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -111,22 +125,31 @@ function(target_precompiled_header) # target [...] header
 			endif()
 		endif()
 
-		add_dependencies(${target} ${pch_target})
```
Are you asking me? I can't tell you if I wanted to. GitHub doesn't provide any context. There's no links to click on to see the files, etc. It's really a poorly designed website unfortunately.

mick-p1982 replied:
I looked at it. Of course, I can't tell you why, unless it was just a stylistic choice. Given how it's not, I wouldn't mess with it. I'm sure I stuck it where it makes logical sense to put it anyway. I can't remember what I was working on a year ago, much less an unplanned diversion. Sorry.

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -111,22 +125,31 @@ function(target_precompiled_header) # target [...] header
+				"/Yu\"${win_header}\""
+				"/Fp\"${win_pch}\""
+				"/FI\"${win_header}\""
+				##"/Fd\"${pdb_dir}\\\\\""
```
It should be removed, unless I'm mistaken about what's in this PR. Maybe it's not up-to-date? I'm a little confused... here ( https://github.com/mick-p1982/CMakePCHCompiler/blob/master/CMakePCHCompiler.cmake ) is the file as it should be. Is that not what this PR is?

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -111,22 +125,31 @@ function(target_precompiled_header) # target [...] header
+			#Note, this adds debug/release/etc. onto the end.
+			set_target_properties(${target} PROPERTIES
+				COMPILE_PDB_OUTPUT_DIRECTORY ${pdb_dir}) ##/Fd\"${pdb_dir}\\\\\"
+			#Careful: set_target_properties is destructive
```
I explained this to @gjasny ... basically don't use `set_target_properties` in a module, since it messes up the user's script.

mick-p1982 replied:
On second thought, I'm not sure what this means. My best guess is it refers to `set(exclude "/Yu${win_header}")` since that is the line under the comments. I think it is adding a variable in order to repair things later on.... where probably `set_target_properties` is used with `exclude`.

mick-p1982 commented on CMakePCHCompiler.cmake:
```diff
@@ -96,7 +96,7 @@ function(target_precompiled_header) # target [...] header
 				# don't forget the slash on the end. (How did this ever work?)
 				# /FI - force include of precompiled header
-				set(flags "/Yc\"${win_header}\" /Fp\"${win_pch}\" /FI\"${win_header}\"") # /Fd\"${pdb_dir}\\\\\"
+				set(flags "/Yc\"${win_header}\" /Fp\"${win_pch}\" /FI\"${win_header}\" /Fd\"${pdb_dir}\\\\\"")
```
My file has `set(flags "/Yc\"${win_header}\" /Fp\"${win_pch}\" /FI\"${win_header}\"")`

nanoant commented:
@mick-p1982 I have cleaned up your changes and squashed & split into atomic (IMHO) commits and I am ready to merge them. (See commits above) Right now, all I need is your review of these changes -- both from the source code content and and commit message wording point of view. Also I would appreciate if you can test them, you change check-out this PR into your local repo as described here: https://help.github.com/articles/checking-out-pull-requests-locally/

@gjasny Can you also please have a look at this PR. I do appreciate your input, especially right now on vacation I am really limited in terms of the testing machine access. Thank you in advance.

nanoant commented:
@mick-p1982 I have cleaned up your changes and squashed & split into atomic (IMHO) commits and I am ready to merge them. (See commits above) Right now, all I need is your review of these changes -- both from the source code content and and commit message wording point of view. Also I would appreciate if you can test them, you change check-out this PR into your local repo as described here: https://help.github.com/articles/checking-out-pull-requests-locally/

@gjasny Can you also please have a look at this PR. I do appreciate your input, especially right now on vacation I am really limited in terms of the testing machine access. Thank you in advance.

mick-p1982 commented:
@gjasny mentioned doing tests. I will definitely make an effort to swap out the main file here the next time I work with this CMake module (I mainly use it personally for testing work with GCC prior to publishing.) I don't think trying to force an artificial test is good for me. And it's not convenient to do a quick test.

FYI: My feeling about PDB is that it's probably not worth maintaining as an option, because it simply doesn't work with Visual Studio if you use the normal ways of building programs with it. So it would only be useful for publishing the PDB files. And I think that if you are that invested into PDB then it probably makes sense to just use Visual Studio to do that. (I think CMake is for publishing to the needs of unknown users. It's a compromise in other words. So the purpose of this module is mainly so you can publish sources that end-users can build/see for themselves quickly. Secondary to that is having a basic development environment out of the box. But that's not a substitute for a proper development environment... which you'll want if you're spending significant time with a project.)

mick-p1982 commented:
@nanoant, until there is a live file, I don't think I can locate a copy of the final file via GitHub, so if you can attache one, I can do a diff with my working-copy. Otherwise I can't do a real review. 

My main advise is to just don't leave anything out, because everything that is changed is changed for a very real reason, e.g. things don't work without the change. I assume the final is more or less what I submitted, or what is sitting in my fork :)

nanoant commented:
@mick-p1982 @gjasny I have tested it with CMake 3.0.2 and Apple Clang, it seems to work fine. Merging. Thanks!

mick-p1982 commented:
Glad to see the public file is updated. With some creativity I checked the RAW button, and copy-pasted it into a new file, for the purpose of diff'ing it with my working copy. (Of course GitHub doesn't provide a download button, why would it?)

Here are some missing lines, but I did not add these, so I assume you (@nanoant) know what these do, and that they are not required...

`set(CMAKE_INCLUDE_SYSTEM_FLAG_${lang}PCH ${CMAKE_INCLUDE_SYSTEM_FLAG_${lang}})`

```
if(CMAKE_VERSION VERSION_LESS 3.12)
			include(Platform/${CMAKE_SYSTEM_NAME}-${CMAKE_${lang}_COMPILER_ID}-${lang}PCH
				OPTIONAL
				)
		else()
			include(Platform/${CMAKE_EFFECTIVE_SYSTEM_NAME}-${CMAKE_${lang}_COMPILER_ID}-${lang}PCH
				OPTIONAL
				)
		endif()
```
-->
```
include(Platform/${CMAKE_SYSTEM_NAME}-${CMAKE_${lang}_COMPILER_ID}-${lang}PCH
			OPTIONAL
			)
```

Looks good to me. I hope it was not too much work. I apologize for not being of more help. My ability with Git is limited to what words to say instead of Checkout/Commit ... and I'm not even sure about the latter! I wish that poking around in confounding mazes of consoles/boxes/buttons were as effortless as talking :)

nanoant commented:
@mick-p1982 No worries, I am happy we have made it through.

As for Git, I can really recommend you reading O'Reilly's [Version Control with Git, 2nd Edition](http://shop.oreilly.com/product/0636920022862.do). I it is really thorough and not only describes how to use Git, but precisely how it works.
